File: engine/scorer.py
"""
SICHER — Safety Scoring Algorithm (Python Implementation)

Scores walking routes by cross-referencing sampled coordinates against
a spatial grid of simulated safety data (lighting, activity, CCTV, emergency).

Uses numpy for efficient array math and a hash-based spatial index for O(1) lookups.
"""

import math
import json
from pathlib import Path
from typing import Optional

try:
    import numpy as np
    HAS_NUMPY = True
except ImportError:
    HAS_NUMPY = False

# Default scoring weights
DEFAULT_WEIGHTS = {
    "lighting": 3,
    "activity": 2,
    "cctv": 2,
    "emergency_proximity": 1,
}


class SafetyGrid:
    """Spatial index over safety grid cells for O(1) coordinate lookups."""

    def __init__(self, grid_data: dict):
        self.grid_size = grid_data.get("grid_size", 0.002)
        self.weights = grid_data.get("weights", DEFAULT_WEIGHTS)
        self.cells = grid_data.get("cells", [])
        self._index: dict = {}
        self._build_index()

    def _build_index(self):
        """Build hash-based spatial index from grid cells."""
        precision = 1000  # ~111m resolution
        for cell in self.cells:
            key = (round(cell["lat"] * precision), round(cell["lng"] * precision))
            self._index[key] = cell

    def find_nearest(self, lat: float, lng: float, max_dist_km: float = 0.5):
        """
        Find the nearest grid cell to a coordinate.

        Args:
            lat: Latitude
            lng: Longitude
            max_dist_km: Maximum search radius in km

        Returns:
            dict or None: Nearest cell data, or None if none within radius
        """
        precision = 1000
        base_lat = round(lat * precision)
        base_lng = round(lng * precision)
        search_radius = 3

        best_cell = None
        best_dist = float("inf")

        for d_lat in range(-search_radius, search_radius + 1):
            for d_lng in range(-search_radius, search_radius + 1):
                key = (base_lat + d_lat, base_lng + d_lng)
                cell = self._index.get(key)
                if cell:
                    dist = haversine(lat, lng, cell["lat"], cell["lng"])
                    if dist < best_dist:
                        best_dist = dist
                        best_cell = cell

        return best_cell if best_dist <= max_dist_km else None


def haversine(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """Calculate haversine distance between two points in km."""
    R = 6371  # Earth radius km
    d_lat = math.radians(lat2 - lat1)
    d_lng = math.radians(lng2 - lng1)
    a = (math.sin(d_lat / 2) ** 2 +
         math.cos(math.radians(lat1)) * math.cos(math.radians(lat2)) *
         math.sin(d_lng / 2) ** 2)
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def sample_route_points(coordinates: list, interval_km: float = 0.05) -> list:
    """
    Sample points along a route at regular intervals (~50m).

    Args:
        coordinates: List of [lng, lat] pairs
        interval_km: Sampling interval in km

    Returns:
        List of sampled [lng, lat] pairs
    """
    if not coordinates:
        return []

    sampled = [coordinates[0]]
    accumulated = 0.0

    for i in range(1, len(coordinates)):
        lng1, lat1 = coordinates[i - 1]
        lng2, lat2 = coordinates[i]
        seg_dist = haversine(lat1, lng1, lat2, lng2)
        accumulated += seg_dist

        if accumulated >= interval_km:
            sampled.append(coordinates[i])
            accumulated = 0.0

    # Always include the last point
    if sampled[-1] != coordinates[-1]:
        sampled.append(coordinates[-1])

    return sampled


def score_route(
    coordinates: list,
    grid: SafetyGrid,
    weights: Optional[dict] = None,
) -> dict:
    """
    Score a single route based on safety grid data.

    Args:
        coordinates: List of [lng, lat] pairs
        grid: SafetyGrid instance
        weights: Optional weight overrides

    Returns:
        dict with score, reason, segments, warning, details
    """
    w = weights or grid.weights
    total_weight = (
        w["lighting"] + w["activity"] + w["cctv"] + w["emergency_proximity"]
    )

    sampled = sample_route_points(coordinates)
    segments = []
    scores = []
    unscored = 0

    # Component accumulators
    total_lighting = 0.0
    total_activity = 0.0
    total_cctv = 0.0
    total_emergency = 0.0
    scored_count = 0

    for lng, lat in sampled:
        cell = grid.find_nearest(lat, lng)
        if cell:
            point_score = (
                w["lighting"] * cell["lighting"]
                + w["activity"] * cell["activity"]
                + w["cctv"] * cell["cctv"]
                + w["emergency_proximity"] * cell["emergency_proximity"]
            ) / total_weight

            scores.append(point_score)
            total_lighting += cell["lighting"]
            total_activity += cell["activity"]
            total_cctv += cell["cctv"]
            total_emergency += cell["emergency_proximity"]
            scored_count += 1

            segments.append({"from_coord": [lng, lat], "score": round(point_score * 10)})
        else:
            unscored += 1
            scores.append(3.0)  # Cautious default
            segments.append({"from_coord": [lng, lat], "score": 30})

    # Calculate final score
    if HAS_NUMPY:
        avg_score = float(np.mean(scores)) if scores else 0.0
    else:
        avg_score = sum(scores) / len(scores) if scores else 0.0

    final_score = round(avg_score * 10)
    final_score = max(0, min(100, final_score))

    # Component averages
    safe_div = max(scored_count, 1)
    avg_l = round(total_lighting / safe_div, 1)
    avg_a = round(total_activity / safe_div, 1)
    avg_c = round(total_cctv / safe_div, 1)
    avg_e = round(total_emergency / safe_div, 1)

    reason = generate_reason(final_score, avg_l, avg_a, avg_c, avg_e, unscored)

    return {
        "score": final_score,
        "reason": reason,
        "segments": segments,
        "warning": final_score < 30,
        "details": {
            "scored_points": scored_count + unscored,
            "unscored_points": unscored,
            "coverage_pct": round((scored_count / max(scored_count + unscored, 1)) * 100),
            "avg_lighting": avg_l,
            "avg_activity": avg_a,
            "avg_cctv": avg_c,
            "avg_emergency": avg_e,
        },
    }


def generate_reason(
    score: int,
    lighting: float,
    activity: float,
    cctv: float,
    emergency: float,
    unscored: int,
) -> str:
    """Generate human-readable safety assessment."""
    parts = []

    if score >= 70:
        if lighting >= 7:
            parts.append("well-lit streets")
        if activity >= 7:
            parts.append("high pedestrian activity")
        if cctv >= 7:
            parts.append("good CCTV coverage")
        if emergency >= 7:
            parts.append("near emergency services")
        if not parts:
            parts.append("generally safe conditions")
        return f"Recommended: {', '.join(parts)}."

    if score >= 40:
        if lighting < 5:
            parts.append("some poorly lit sections")
        if activity < 5:
            parts.append("low foot traffic in parts")
        if cctv < 5:
            parts.append("limited surveillance")
        if not parts:
            parts.append("moderate safety conditions")
        return f"Caution: {', '.join(parts)}."

    # Low score
    if lighting < 4:
        parts.append("poorly lit area")
    if activity < 3:
        parts.append("minimal foot traffic")
    if cctv < 3:
        parts.append("no surveillance coverage")
    if unscored > 3:
        parts.append("unmapped zones along route")
    if not parts:
        parts.append("generally unsafe conditions")
    return f"Warning: {', '.join(parts)}. Consider a ride service instead."


def load_grid_from_file(filepath: str = None) -> SafetyGrid:
    """Load safety grid from JSON file."""
    if filepath is None:
        filepath = str(
            Path(__file__).parent.parent / "backend" / "data" / "safety_grid.json"
        )
    with open(filepath, "r") as f:
        data = json.load(f)
    return SafetyGrid(data)

File: backend/lib/scorer.js
/**
 * SICHER Safety Scoring Engine (JavaScript implementation)
 * 
 * Scores walking routes by cross-referencing coordinates against
 * a simulated safety grid. Uses weighted averaging of:
 *   - Lighting (weight: 3)
 *   - Pedestrian Activity (weight: 2)
 *   - CCTV Coverage (weight: 2)
 *   - Emergency Service Proximity (weight: 1)
 */

const path = require('path');
const fs = require('fs');

// Load and index the safety grid
let gridData = null;
let gridIndex = null;

/**
 * Load the safety grid from JSON file and build a spatial index.
 * Uses a simple hash-based grid for O(1) lookups.
 */
function loadGrid() {
  const gridPath = path.join(__dirname, '..', 'data', 'safety_grid.json');
  const raw = fs.readFileSync(gridPath, 'utf-8');
  gridData = JSON.parse(raw);

  // Build spatial hash index for fast lookups
  gridIndex = new Map();
  const precision = 1000; // ~111m resolution

  for (const cell of gridData.cells) {
    const key = `${Math.round(cell.lat * precision)},${Math.round(cell.lng * precision)}`;
    gridIndex.set(key, cell);
  }

  console.log(`[SICHER] Safety grid loaded: ${gridData.cells.length} cells indexed`);
  return gridData;
}

/**
 * Find the nearest grid cell to a coordinate.
 * Uses spatial hash with expanding search radius.
 * 
 * @param {number} lat - Latitude
 * @param {number} lng - Longitude
 * @returns {object|null} Nearest grid cell or null if too far
 */
function findNearestCell(lat, lng) {
  if (!gridIndex) loadGrid();

  const precision = 1000;
  const searchRadius = 3; // search ±3 hash cells

  let bestCell = null;
  let bestDist = Infinity;

  const baseLat = Math.round(lat * precision);
  const baseLng = Math.round(lng * precision);

  for (let dLat = -searchRadius; dLat <= searchRadius; dLat++) {
    for (let dLng = -searchRadius; dLng <= searchRadius; dLng++) {
      const key = `${baseLat + dLat},${baseLng + dLng}`;
      const cell = gridIndex.get(key);
      if (cell) {
        const dist = haversineDistance(lat, lng, cell.lat, cell.lng);
        if (dist < bestDist) {
          bestDist = dist;
          bestCell = cell;
        }
      }
    }
  }

  // Only return cells within 500m
  if (bestDist > 0.5) return null;
  return bestCell;
}

/**
 * Calculate haversine distance between two coordinates in km.
 * 
 * @param {number} lat1 - Latitude of point 1
 * @param {number} lng1 - Longitude of point 1
 * @param {number} lat2 - Latitude of point 2
 * @param {number} lng2 - Longitude of point 2
 * @returns {number} Distance in kilometers
 */
function haversineDistance(lat1, lng1, lat2, lng2) {
  const R = 6371; // Earth radius in km
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

function toRad(deg) {
  return (deg * Math.PI) / 180;
}

/**
 * Sample points along a route at regular intervals (~50m).
 * 
 * @param {Array} coordinates - Array of [lng, lat] pairs
 * @returns {Array} Sampled [lng, lat] pairs
 */
function sampleRoutePoints(coordinates) {
  if (!coordinates || coordinates.length === 0) return [];
  
  const sampleDistKm = 0.05; // 50 meters
  const sampled = [coordinates[0]];
  let accumulated = 0;

  for (let i = 1; i < coordinates.length; i++) {
    const [lng1, lat1] = coordinates[i - 1];
    const [lng2, lat2] = coordinates[i];
    const segDist = haversineDistance(lat1, lng1, lat2, lng2);
    accumulated += segDist;

    if (accumulated >= sampleDistKm) {
      sampled.push(coordinates[i]);
      accumulated = 0;
    }
  }

  // Always include last point
  if (sampled[sampled.length - 1] !== coordinates[coordinates.length - 1]) {
    sampled.push(coordinates[coordinates.length - 1]);
  }

  return sampled;
}

/**
 * Score a single route based on safety grid data.
 * 
 * @param {Array} coordinates - Array of [lng, lat] pairs
 * @param {object} [weights] - Optional weight overrides
 * @returns {Promise<object>} Score result { score, reason, segments, warning }
 */
async function scoreRoute(coordinates, weights = null) {
  if (!gridData) loadGrid();

  const w = weights || gridData.weights;
  const totalWeight = w.lighting + w.activity + w.cctv + w.emergency_proximity;

  const sampledPoints = sampleRoutePoints(coordinates);
  const segments = [];
  let totalScore = 0;
  let scoredPoints = 0;
  let unscoredPoints = 0;

  // Track component averages for reason generation
  let totalLighting = 0;
  let totalActivity = 0;
  let totalCctv = 0;
  let totalEmergency = 0;

  for (const [lng, lat] of sampledPoints) {
    const cell = findNearestCell(lat, lng);
    if (cell) {
      const pointScore =
        (w.lighting * cell.lighting +
          w.activity * cell.activity +
          w.cctv * cell.cctv +
          w.emergency_proximity * cell.emergency_proximity) /
        totalWeight;

      totalScore += pointScore;
      totalLighting += cell.lighting;
      totalActivity += cell.activity;
      totalCctv += cell.cctv;
      totalEmergency += cell.emergency_proximity;
      scoredPoints++;

      segments.push({
        from: [lng, lat],
        score: Math.round(pointScore * 10),
      });
    } else {
      unscoredPoints++;
      // Assign a cautious default score of 3 for unscored areas
      totalScore += 3;
      scoredPoints++;
      segments.push({
        from: [lng, lat],
        score: 30,
      });
    }
  }

  // Calculate final normalized score (0-100)
  const avgScore = scoredPoints > 0 ? totalScore / scoredPoints : 0;
  const finalScore = Math.round(avgScore * 10);

  // Calculate component averages
  const realScored = scoredPoints - unscoredPoints;
  const avgLighting = realScored > 0 ? (totalLighting / realScored).toFixed(1) : 0;
  const avgActivity = realScored > 0 ? (totalActivity / realScored).toFixed(1) : 0;
  const avgCctv = realScored > 0 ? (totalCctv / realScored).toFixed(1) : 0;
  const avgEmergency = realScored > 0 ? (totalEmergency / realScored).toFixed(1) : 0;

  // Generate human-readable reason — use AI if key available, otherwise fallback
  let reason;
  const geminiKey = process.env.GOOGLE_GEMINI_API_KEY || process.env.GOOGLE_MAPS_API_KEY;
  if (geminiKey && geminiKey !== 'your_google_maps_key_here') {
    reason = await generateAIReason(finalScore, {
      lighting: avgLighting,
      activity: avgActivity,
      cctv: avgCctv,
      emergency: avgEmergency,
      unscored: unscoredPoints
    }, geminiKey);
  } else {
    reason = generateReason(finalScore, avgLighting, avgActivity, avgCctv, avgEmergency, unscoredPoints);
  }

  return {
    score: finalScore,
    reason,
    segments,
    warning: finalScore < 30,
    details: {
      scored_points: scoredPoints,
      unscored_points: unscoredPoints,
      coverage_pct: Math.round((realScored / (scoredPoints || 1)) * 100),
      avg_lighting: parseFloat(avgLighting),
      avg_activity: parseFloat(avgActivity),
      avg_cctv: parseFloat(avgCctv),
      avg_emergency: parseFloat(avgEmergency),
    },
  };
}

/**
 * Generate an AI-powered safety assessment using Google Gemini.
 */
async function generateAIReason(score, factors, apiKey) {
  const fetch = require('node-fetch');
  const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key=${apiKey}`;

  const prompt = `
    As a neighborhood safety assistant, analyze this walking route:
    - Safety Score: ${score}/100
    - Lighting: ${factors.lighting}/10 (0=dark, 10=bright)
    - Activity: ${factors.activity}/10 (0=isolated, 10=busy)
    - CCTV: ${factors.cctv}/10
    - Emergency Services: ${factors.emergency}/10
    - Unmapped regions: ${factors.unscored} points
    
    Provide a concise, professional 1-2 sentence safety assessment for a traveler. 
    If score < 40, recommend a taxi. If score > 70, mention why it's safe.
    Don't use lists. Just natural speech.
  `;

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        contents: [{ parts: [{ text: prompt }] }]
      })
    });
    const data = await response.json();
    return data.candidates?.[0]?.content?.parts?.[0]?.text?.trim() || generateReason(score, factors.lighting, factors.activity, factors.cctv, factors.emergency, factors.unscored);
  } catch (err) {
    console.error('[SICHER] Gemini AI error:', err.message);
    return generateReason(score, factors.lighting, factors.activity, factors.cctv, factors.emergency, factors.unscored);
  }
}

/**
 * Generate a human-readable safety assessment.
 * 
 * @param {number} score - Final safety score (0-100)
 * @param {number} lighting - Average lighting score
 * @param {number} activity - Average activity score
 * @param {number} cctv - Average CCTV score
 * @param {number} emergency - Average emergency proximity score
 * @param {number} unscored - Number of unscored points
 * @returns {string} Natural language reason
 */
function generateReason(score, lighting, activity, cctv, emergency, unscored) {
  const parts = [];

  if (score >= 70) {
    if (lighting >= 7) parts.push('well-lit streets');
    if (activity >= 7) parts.push('high pedestrian activity');
    if (cctv >= 7) parts.push('good CCTV coverage');
    if (emergency >= 7) parts.push('near emergency services');
    if (parts.length === 0) parts.push('generally safe conditions');
    return `Recommended: ${parts.join(', ')}.`;
  }

  if (score >= 40) {
    if (lighting < 5) parts.push('some poorly lit sections');
    if (activity < 5) parts.push('low foot traffic in parts');
    if (cctv < 5) parts.push('limited surveillance');
    if (parts.length === 0) parts.push('moderate safety conditions');
    return `Caution: ${parts.join(', ')}.`;
  }

  // Low score
  if (lighting < 4) parts.push('poorly lit area');
  if (activity < 3) parts.push('minimal foot traffic');
  if (cctv < 3) parts.push('no surveillance coverage');
  if (unscored > 3) parts.push('unmapped zones along route');
  if (parts.length === 0) parts.push('generally unsafe conditions');
  return `Warning: ${parts.join(', ')}. Consider a ride service instead.`;
}

/**
 * Calculate route distance in km from coordinate array.
 * 
 * @param {Array} coordinates - Array of [lng, lat] pairs
 * @returns {number} Total distance in kilometers
 */
function calculateDistanceKm(coordinates) {
  let total = 0;
  for (let i = 1; i < coordinates.length; i++) {
    const [lng1, lat1] = coordinates[i - 1];
    const [lng2, lat2] = coordinates[i];
    total += haversineDistance(lat1, lng1, lat2, lng2);
  }
  return Math.round(total * 100) / 100;
}

/**
 * Estimate walking duration in minutes (avg 5 km/h).
 * 
 * @param {number} distanceKm - Distance in kilometers
 * @returns {number} Estimated minutes
 */
function estimateWalkingMinutes(distanceKm) {
  return Math.round((distanceKm / 5) * 60);
}

// Pre-load grid on module import
try {
  loadGrid();
} catch (err) {
  console.warn('[SICHER] Could not pre-load safety grid:', err.message);
}

module.exports = {
  scoreRoute,
  calculateDistanceKm,
  estimateWalkingMinutes,
  findNearestCell,
  haversineDistance,
  sampleRoutePoints,
  loadGrid,
};
